consensus/safety-rules/src/safety_rules.rs

```diff
@@ -126,13 +134,13 @@ impl SafetyRules {
         &self.validator_signer
     }
 
-    /// Learn about a new quorum certificate. Several things can happen as a result of that:
-    /// 1) update the preferred block to a higher value.
-    /// 2) commit some blocks.
-    /// In case of commits the last committed block is returned.
-    /// Requires that all the ancestors of the block are available for at least up to the last
-    /// committed block, might panic otherwise.
-    /// The update function is invoked whenever a system learns about a potentially high QC.
+    /// Learn about a new quorum certificate. In normal state, this updates the preferred round,
+    /// if the parent is greater than our current preferred round. This can also trigger upgrading
+    /// to a new epoch.
+    /// @TODO verify signatures of the QC
+    /// @TODO return back an error if the QC has an older preferred round, signatures are
```

Comment:
older preferred round is not an error probably, we pass every QC we received to safety rules

Reply:
Perhaps we should just return back an enum rather than a result since passing in an old QC should be a no-op and anything that doesn't increase the state of safety rules could probably be ignored by the rest of the system unless it is part of state synchronization.